### PR TITLE
[CHORE] [Integration test] Create test cases for Login page

### PR DESCRIPTION
**Pull request by manh-t**
## What happened 👀

Write UI test for Login page

:yeah:

## Proof Of Work 📹

https://user-images.githubusercontent.com/60863885/126606343-29bccae4-de4e-4e30-8aeb-93bc9d269fac.mp4

```json
fvm flutter drive --driver=test_driver/integration_test.dart  --target=integration_test/login_test.dart --flavor staging
Running "flutter pub get" in flutter_survey...                   1,075ms
Running Gradle task 'assembleStagingDebug'...                       5.6s
✓  Built build/app/outputs/flutter-apk/app-staging-debug.apk.
Installing build/app/outputs/flutter-apk/app.apk...                 3.6s
I/OpenGLRenderer(25397): Davey! duration=1004ms; Flags=1, IntendedVsync=22896485486453, Vsync=22897335486419, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=22897339122518, AnimationStart=22897339273518, PerformTraversalsStart=22897341542518, DrawStart=22897352521518, SyncQueued=22897367158518, SyncStart=22897369356518, IssueDrawCommandsStart=22897370840518, SwapBuffers=22897433011518, FrameCompleted=22897491953518, DequeueBufferDuration=30232000, QueueBufferDuration=740000, 
VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:56156/1DoZKkJHRn0=/
VMServiceFlutterDriver: Isolate found with number: 2527284513716931
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
D/EGL_emulation(25397): eglMakeCurrent: 0xea25f8c0: ver 3 0 (tinfo 0xe875fa00)
D/eglCodecCommon(25397): setVertexArrayObject: set vao to 0 (0) 1 0
I/flutter (25397): 00:00 +0: Login Page When login with invalid email or password, it returns error message
VMServiceFlutterDriver: Connected to Flutter application.
I/flutter (25397): 00:10 +1: Login Page When login with valid email and password, it returns DashboardPage
I/flutter (25397): 00:18 +1: Login Page When login with invalid email or password, it returns error message
I/flutter (25397): (The following exception is now available via WidgetTester.takeException:)
I/flutter (25397): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter (25397): The following ProviderException was thrown building Consumer(dirty, dependencies:
I/flutter (25397): [UncontrolledProviderScope], state: _ConsumerState#eba1c):
I/flutter (25397): An exception was thrown while building AutoDisposeStateNotifierProvider<DashboardModel,
I/flutter (25397): DashboardState>#00005().
I/flutter (25397): 
I/flutter (25397): Thrown exception:
I/flutter (25397): An exception was thrown while building _AutoDisposeNotifierProvider<DashboardModel>#00006().
I/flutter (25397): 
I/flutter (25397): Thrown exception:
I/flutter (25397): UnimplementedError
I/flutter (25397): 
I/flutter (25397): Stack trace:
I/flutter (25397): #0      FakeUserService.getUserProfile
I/flutter (25397): (package:flutter_survey/api/mock_service/fake_user_service.dart:9:5)
I/flutter (25397): #1      UserRepositoryImpl.getUserProfile
I/flutter (25397): (package:flutter_survey/api/repository/user_repository.dart:18:10)
I/flutter (25397): #2      GetUserProfileUseCase.call
I/flutter (25397): (package:flutter_survey/usecase/get_user_profile_use_case.dart:15:28)
I/flutter (25397): #3      DashboardModel.init (package:flutter_survey/pages/dashboard/dashboard_model.dart:39:30)
I/flutter (25397): #4      new DashboardModel (package:flutter_survey/pages/dashboard/dashboard_model.dart:24:5)
I/flutter (25397): #5      dashboardModelProvider.<anonymous closure>
I/flutter (25397): (package:flutter_survey/pages/dashboard/dashboard_page.dart:17:10)
...
I/flutter (25397): #133    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4599:16)
I/flutter (25397): #134    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4746:11)
I/flutter (25397): 00:19 +2: (tearDownAll)
I/flutter (25397): 00:19 +3: All tests passed!
All tests passed.
```

### CODE REVIEW

File: lib/api/mock_service/fake_oauth_service.dart [hunk 1]
```diff
@@ -7,10 +9,20 @@ import 'package:flutter_survey/api/service/oauth_service.dart';
 
 class FakeOauthService extends BaseOauthService {
```
Comment (markgravity) [inline lib/api/mock_service/fake_oauth_service.dart:10]: Can we use mockito or you're trying to don't use it in this project? 🤔

File: integration_test/login_test.dart [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    late Finder emailField;
+    late Finder passwordField;
+    late Finder submitBtn;
+
+    setUp(() {
+      app.appEnvironment = Environment.test;
+      app.main();
+      emailField = find.byKey(Key('tfLoginEmail'));
+      passwordField = find.byKey(Key('tfLoginPassword'));
+      submitBtn = find.byKey(Key('btnLogin'));
```
Comment (markgravity) [inline integration_test/login_test.dart:26]: I think we should use constant here. What do you think? 🤔

Comment (markgravity) [inline integration_test/login_test.dart:26]: I think we should use constant here. What do you think? 🤔

File: lib/api/mock_service/fake_oauth_service.dart (hunk 1 above)
Comment (manh-t) [reply lib/api/mock_service/fake_oauth_service.dart:10]: @markgravity For Flutter 2.2, I'm having an issue that can't use `mockito` to mock the response, so I go this way and will get back to resolve the issue when having time. 🙏

**Comment by manh-t** [comment]
@markgravity updated 🙏

**Review by markgravity — APPROVED** [review]

File: integration_test/fake_services/fake_oauth_service.dart [hunk 3]
```diff
@@ -1,4 +1,27 @@
...
 import 'package:flutter_test/flutter_test.dart';
 
-class FakeOAuthService extends Fake implements BaseOAuthService {}
+import 'fake_data.dart';
+
+class FakeOAuthService extends Fake implements BaseOAuthService {
+  @override
+  Future<BaseHttpResponse<OAuthTokenResponse>> login(
+      OAuthLoginRequest body) async {
+    await Future.delayed(Duration(seconds: 2));
```
Comment (sleepylee) [inline integration_test/fake_services/fake_oauth_service.dart:14]: I'd recommend to low down this time, usually, the data is instant, so you don't have to wait for too long to complete the test 😄

File: test_resources/oauth_login.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+        "type": "token",
+        "attributes": {
+            "access_token": "lbxD2K2BjbYtNzz8xjvh2FvSKx838KBCf79q773kq2c",
+            "token_type": "Bearer",
+            "expires_in": 7200,
+            "refresh_token": "3zJz2oW0njxlj_I3ghyUBF7ZfdQKYXd2n0ODlMkAjHc",
+            "created_at": 1597169495
+        }
+    }
+}
```
Comment (sleepylee) [inline test_resources/oauth_login.json:13]: add eof line

**Review by sleepylee — APPROVED** [review]
minor comments

File: integration_test/fake_services/fake_oauth_service.dart (hunk 3 above)
Comment (manh-t) [reply integration_test/fake_services/fake_oauth_service.dart:14]: sure, will decrease the duration to `300` maybe :D

**Pull request merged**